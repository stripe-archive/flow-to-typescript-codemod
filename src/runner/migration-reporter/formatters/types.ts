import { MigrationReport } from '..'

export type MigrationReportFormatter = (
  report: MigrationReport
) => Promise<void>
