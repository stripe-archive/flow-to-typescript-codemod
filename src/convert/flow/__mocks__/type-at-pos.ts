export const flowTypeAtPos = jest.fn()
