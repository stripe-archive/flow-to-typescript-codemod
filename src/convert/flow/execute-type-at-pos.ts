import childProcess from "child_process";
import * as t from "@babel/types";

/**
 * Actually executes `flow type-at-pos`. This will be called behind a throttle.
 */
export async function executeFlowTypeAtPos(
  filePath: string,
  location: t.SourceLocation,
  yarnPath?: string
): Promise<string> {
  const { line, column } = location.start;
  const command = `${
    yarnPath ? yarnPath : "$(yarn bin)/flow"
  } type-at-pos "${filePath}" ${line} ${
    column + 1
  } --json --from "typescriptify" --quiet`;

  // Actually run Flow...
  const stdout = await new Promise<string>((resolve, reject) => {
    childProcess.exec(command, (error, stdout) => {
      if (error) {
        reject(error);
      } else {
        resolve(stdout);
      }
    });
  });
  return stdout;
}
