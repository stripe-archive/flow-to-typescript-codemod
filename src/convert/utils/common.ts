import * as t from '@babel/types'
import traverse, { NodePath, Scope } from '@babel/traverse'
import { types } from 'recast'
import { TransformerInput } from '../transformer'
import MigrationReporter from '../../runner/migration-reporter'
import { logger } from '../../runner/logger'

/**
 * Determine whether the file contains any JSX
 * @param file : File source to check
 */
export function hasJSX({ file }: TransformerInput): boolean {
  let found = false
  traverse(file, {
    JSXElement() {
      found = true
    },
    JSXFragment() {
      found = true
    },
  })

  return found
}

/**
 * Determine whether function block returns null
 * Helpful when typing React functional components
 */
export function hasNullReturn(
  body: t.BlockStatement,
  scope: Scope | undefined,
  parentPath: NodePath<t.Node> | null | undefined
): boolean {
  let found = false
  traverse(
    body,
    {
      ReturnStatement(path) {
        if (path.node.argument?.type === 'NullLiteral') {
          found = true
        }
      },
    },
    scope,
    parentPath
  )

  return found
}

/**
 * Determine whether the file contains any type declarations
 * @param file : File source to check
 */
export function hasDeclaration(file: t.File): boolean {
  let found = false
  traverse(file, {
    DeclareModule() {
      found = true
    },
    DeclareExportDeclaration() {
      found = true
    },
    DeclareClass() {
      found = true
    },
  })

  return found
}

/**
 * Construct a TS type identifier.
 * @param name
 * @param optional
 * @param typeAnnotation
 * @returns
 */
export function buildTSIdentifier(
  name: string,
  optional?: boolean | null,
  typeAnnotation?: t.TSTypeAnnotation | null
): t.Identifier {
  const identifier = t.identifier(name)
  if (optional != null) identifier.optional = optional
  if (typeAnnotation != null) identifier.typeAnnotation = typeAnnotation
  return identifier
}

/**
 * Is this a literal expression? Includes literal objects and functions.
 */
export function isComplexLiteral(expression: t.Expression): boolean {
  if (t.isLiteral(expression)) {
    return true
  }
  if (expression.type === 'Identifier' && expression.name === 'undefined') {
    return true
  }

  if (expression.type === 'ArrayExpression') {
    for (const element of expression.elements) {
      if (element === null) {
        continue
      }
      if (element.type === 'SpreadElement') {
        if (!isComplexLiteral(element.argument)) {
          return false
        } else {
          continue
        }
      }
      if (!isComplexLiteral(element)) {
        return false
      }
    }
    return true
  }

  if (expression.type === 'ObjectExpression') {
    for (const property of expression.properties) {
      if (property.type === 'ObjectMethod') {
        return false
      } else if (property.type === 'SpreadElement') {
        return false
      } else {
        if (property.computed && !isComplexLiteral(property.key)) {
          return false
        }
        if (
          t.isExpression(property.value) &&
          !isComplexLiteral(property.value)
        ) {
          return false
        }
      }
    }
    return true
  }

  return false
}

/**
 * Are we inside `createReactClass()`?
 */
export function isInsideCreateReactClass(path: NodePath<t.Node>): boolean {
  if (
    path.node.type === 'CallExpression' &&
    path.node.callee.type === 'Identifier' &&
    path.node.callee.name === 'createReactClass'
  ) {
    return true
  }

  if (path.parentPath) {
    return isInsideCreateReactClass(path.parentPath)
  }

  return false
}

/**
 * Copies the location and comments of one node to a new node.
 */
export function inheritLocAndComments(oldNode: t.Node, newNode: t.Node) {
  newNode.loc = oldNode.loc

  // Recast uses a different format for comments then Babel.
  if ('comments' in oldNode) {
    // @ts-expect-error comments doesn't exist on babel type
    newNode.comments = oldNode.comments
    delete oldNode.comments
  }
}

export function addCommentsAtHeadOfNode(
  rootNode: types.namedTypes.Node | undefined,
  comments: (types.namedTypes.CommentBlock | types.namedTypes.CommentLine)[]
) {
  if (rootNode !== undefined) {
    rootNode.comments = rootNode.comments || []
    rootNode.comments.unshift(...comments)
  } else {
    logger.warn(`Cannot add comments ${comments} to empty node!`)
  }
}

export function addEmptyLineInProgramPath(path: NodePath<t.Program>) {
  path.unshiftContainer('body', t.noop())
}

/**
 * Recast uses a different format for comments. We need to manually copy them over to the new node.
 * We also attach the old location so that Recast prints it at the same place.
 *
 * https://github.com/benjamn/recast/issues/572
 */
export function replaceWith(
  path: NodePath<t.Node>,
  node: t.Node,
  filePath: string,
  reporter: MigrationReporter
) {
  inheritLocAndComments(path.node, node)
  try {
    path.replaceWith(node)
  } catch (e) {
    // Catch the error so conversion of the file can continue.
    reporter.error(filePath, e)
  }
}

/**
 * Tries to return the nearest LOC, and returns a default if not found.
 */
export function getLoc<TNodeType extends t.Node>(
  node: TNodeType
): t.SourceLocation {
  return (
    node.loc ??
    (node as t.FunctionDeclaration).body?.loc ?? {
      start: {
        line: 0,
        column: 0,
      },
      end: {
        line: 0,
        column: 0,
      },
    }
  )
}

export const GlobalTypes = {
  TimeoutID: 'number',
  IntervalID: 'number',
  ImmediateID: 'number',
  immediateID: 'number',
  AnimationFrameID: 'number',
  RequestOptions: 'RequestInit',
} as const

export const LiteralTypes = {
  String: 'string',
  Number: 'number',
  Boolean: 'boolean',
  Symbol: 'symbol',
} as const

export const JEST_MOCK_METHODS = ['mock', 'requireActual']
