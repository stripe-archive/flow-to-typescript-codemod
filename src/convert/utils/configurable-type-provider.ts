import * as t from "@babel/types";

export class ConfigurableTypeProvider {
  readonly useStrictAnyFunctionType: boolean = false;

  readonly useStrictAnyObjectType: boolean = false;

  constructor({
    useStrictAnyFunctionType,
    useStrictAnyObjectType,
  }: {
    useStrictAnyObjectType: boolean;
    useStrictAnyFunctionType: boolean;
  }) {
    this.useStrictAnyObjectType = useStrictAnyObjectType;
    this.useStrictAnyFunctionType = useStrictAnyFunctionType;
  }

  public get flowAnyObjectType(): t.TSType {
    if (this.useStrictAnyObjectType) {
      // Record<any, any>
      return t.tsTypeReference(
        t.identifier("Record"),
        t.tsTypeParameterInstantiation([t.tsAnyKeyword(), t.tsAnyKeyword()])
      );
    } else {
      return t.tsAnyKeyword();
    }
  }

  public get flowAnyFunctionType(): t.TSType {
    if (this.useStrictAnyFunctionType) {
      return t.tsTypeReference(t.identifier("Function"));
    } else {
      return t.tsAnyKeyword();
    }
  }
}
