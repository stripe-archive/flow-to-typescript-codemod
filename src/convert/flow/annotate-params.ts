import * as t from '@babel/types'
import { NodePath } from '@babel/traverse'
import MigrationReporter from '../../runner/migration-reporter'
import { State } from '../../runner/state'
import { flowTypeAtPos } from './type-at-pos'
import { migrateType } from '../migrate/type'

/**
 * There are some cases where it's fine if variables or parameters don't have a type
 * These are often cases where the parent of the node provides a type
 * In these cases we want to avoid annotating a type from Flow, which is often different
 * and just ends up causing more errors. Flow is also slow to query so this helps
 * reduce calls.
 */
function parentProvidesImplicitType(path: NodePath<t.Node>): boolean {
  const parent = path.parentPath
  switch (parent?.node.type) {
    case 'VariableDeclarator':
      return (
        t.isIdentifier(parent.node.id) && parent.node.id.typeAnnotation != null
      )
    case 'CallExpression':
      return (
        t.isIdentifier(parent.node.callee) &&
        parent.node.callee.typeAnnotation != null
      )
    case 'AssignmentExpression':
      return (
        t.isIdentifier(parent.node.left) &&
        parent.node.left.typeAnnotation != null
      )
    case 'TSAsExpression':
    case 'TypeCastExpression':
    case 'ObjectProperty':
    case 'JSXExpressionContainer':
      return true
    case 'ClassProperty':
    case 'ClassPrivateProperty':
      return parent.node.typeAnnotation != null
    case 'ArrowFunctionExpression':
      return parent.node.returnType != null

    // Currently we don’t support type annotations in object properties
    // case 'ObjectProperty':
    // case 'ObjectExpression':
    //   return parentProvidesImplicitType(parent);
  }

  return false
}

/**
 * Adds a type annotation to all unannotated function parameters.
 */
export function annotateParamsWithFlowTypeAtPos(
  reporter: MigrationReporter,
  state: State,
  params: t.ClassMethod['params'],
  path: NodePath,
  isInsideCreateReactClass?: boolean
): Promise<unknown> {
  if (!isInsideCreateReactClass && parentProvidesImplicitType(path)) {
    return Promise.resolve()
  }

  if (state.config.disableFlow) {
    for (const param of params) {
      if (param.type === 'Identifier' && !param.typeAnnotation) {
        param.typeAnnotation = t.tsTypeAnnotation(t.tsUnknownKeyword())
        reporter.disableFlowCheck(state.config.filePath, param.loc!)
      }
    }

    return Promise.resolve()
  }

  const awaitPromises: Array<Promise<void>> = []

  for (const param of params) {
    if (param.type === 'Identifier' && !param.typeAnnotation) {
      awaitPromises.push(
        (async () => {
          // Get the type Flow is inferring for this unannotated function parameter.
          const flowType = await flowTypeAtPos(state, param.loc!, reporter)
          if (flowType === null) return

          // If Flow inferred `empty` then that means there were no calls to the
          // function and therefore no “lower type bounds” for the parameter. This
          // means you can do anything with the type effectively making it any. So
          // treat it as such.
          const tsType =
            flowType.type === 'EmptyTypeAnnotation'
              ? t.tsAnyKeyword()
              : migrateType(reporter, state, flowType)

          // Add the type annotation! Yaay.
          param.typeAnnotation = t.tsTypeAnnotation(tsType)
        })()
      )
    }
  }

  return Promise.all(awaitPromises)
}
