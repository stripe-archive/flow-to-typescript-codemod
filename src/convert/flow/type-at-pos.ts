import * as t from "@babel/types";
import * as recast from "recast";
import * as recastFlowParser from "recast/parsers/flow";
import { executeFlowTypeAtPos } from "./execute-type-at-pos";
import { State } from "../../runner/state";
import MigrationReporter from "../../runner/migration-reporter";
import { transformPrivateTypes } from "../private-types";
/**
 * Runs Flow to get the inferred type at a given position. Uses the Flow server so once the Flow
 * server is running this should be pretty fast. We use this to add explicit annotations where Flow
 * needs some help.
 *
 * Queued so that we don’t overload the Flow server.
 */
export function flowTypeAtPos(
  state: State,
  location: t.SourceLocation,
  migrationReporter: MigrationReporter
): Promise<t.FlowType | null> {
  let resolve: (value: string) => void;
  let reject: (error: unknown) => void;

  const promise = new Promise<string>((_resolve, _reject) => {
    resolve = _resolve;
    reject = _reject;
  });

  flowTypeAtPosQueue.push({
    filePath: state.config.filePath,
    location,
    migrationReporter,
    resolve: resolve!,
    reject: reject!,
  });

  if (processingFlowTypeAtPosQueue === false) {
    processFlowTypeAtPosQueue(state.config.yarnPath);
  }

  return promise
    .then((stdOut) =>
      processFlowTypeAtPosStdout(stdOut, migrationReporter, state, location)
    )
    .catch(() => {
      return null;
    });
}

/**
 * Are we currently processing `flowTypeAtPosQueue`? If so then don’t call
 * `processFlowTypeAtPosQueue()` a second time.
 */
let processingFlowTypeAtPosQueue = false;

/**
 * Holds all the pending `flowTypeAtPos()` calls.
 */
const flowTypeAtPosQueue: Array<{
  filePath: string;
  location: t.SourceLocation;
  migrationReporter: MigrationReporter;
  resolve: (value: string) => void;
  reject: (error: unknown) => void;
}> = [];

/**
 * Continually process all the entries in `flowTypeAtPosQueue`.
 */
function processFlowTypeAtPosQueue(yarnPath?: string) {
  processingFlowTypeAtPosQueue = true;

  const entry = flowTypeAtPosQueue.shift();

  if (!entry) {
    processingFlowTypeAtPosQueue = false;
    return;
  }

  executeFlowTypeAtPos(entry.filePath, entry.location, yarnPath).then(
    (value) => {
      // Start the next asynchronous `flow type-at-pos` request before resolving the entry!
      // When we resolve the entry some synchronous work will be done to parse the result.
      // We can do that work concurrently while `flow type-at-pos` works.
      processFlowTypeAtPosQueue(yarnPath);
      entry.resolve(value);
    },
    (value) => {
      processFlowTypeAtPosQueue(yarnPath);
      entry.migrationReporter.flowFailToParse(
        entry.filePath,
        entry.location,
        value as Error
      );
      entry.reject(value);
    }
  );
}

/**
 * Processes the standard output of `flow type-at-pos`.
 */
function processFlowTypeAtPosStdout(
  stdout: string,
  migrationReporter: MigrationReporter,
  state: State,
  location: t.SourceLocation
): t.FlowType | null {
  // Sanitize stdout...
  // `any(implicit)` -> `any`
  let { type } = JSON.parse(stdout) as { type: string };
  const isExplicit = type.includes("(explicit)");
  type = type.replace("(explicit)", "");
  type = type.replace("(implicit)", "");

  // Flow does not know the type at this location.
  if (type === "unknown") {
    migrationReporter.unknownFlowType(state.config.filePath, location);
    return null;
  }

  if (type === "any" && !isExplicit && !state.config.silenceNonCriticalLogs) {
    migrationReporter.anyFlowType(state.config.filePath, location);
    return null;
  }

  // The inferred Flow type is really big, a human probably would not have written it. Don’t
  // return the type.
  if (type.length >= 100 && !state.config.silenceNonCriticalLogs) {
    migrationReporter.complexFlowType(state.config.filePath, location, type);
    return null;
  }

  try {
    // Parse the Flow type and return it!
    const flowType: t.File = recast.parse(`type T = ${type};`, {
      parser: recastFlowParser,
    });

    // Run our pre-processing step on the types
    transformPrivateTypes({
      file: flowType,
      reporter: migrationReporter,
      state,
    });

    const node = (flowType.program.body[0] as t.TypeAlias).right;

    // `function f(x: string | any)`
    if (
      node.type === "UnionTypeAnnotation" &&
      node.types.some((unionType) => unionType.type === "AnyTypeAnnotation") &&
      !state.config.silenceNonCriticalLogs
    ) {
      migrationReporter.anyFlowType(state.config.filePath, location);
      return null;
    }

    return node;
  } catch (e) {
    migrationReporter.flowFailToParse(
      state.config.filePath,
      location,
      e as Error
    );
    return null;
  }
}
