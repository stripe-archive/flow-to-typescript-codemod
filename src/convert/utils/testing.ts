import * as recast from 'recast'
import * as t from '@babel/types'
import * as recastFlowParser from 'recast/parsers/flow'
import { Project, SourceFile } from 'ts-morph'
import { recastOptions } from '../../runner/process-batch'
import { runTransforms } from '../../runner/run-transforms'
import MigrationReporter from '../../runner/migration-reporter'
import { defaultTransformerChain } from '../default-transformer-chain'
import { watermarkTransformRunner } from '../transform-runners'
import { Transformer } from '../transformer'
import { State } from '../../runner/state'
import { ConfigurableTypeProvider } from './configurable-type-provider'
import { FixCommandCliArgs } from '../../cli/arguments'
import { FixCommandState } from '../../fix/state'

const MockedMigrationReporter =
  MigrationReporter as unknown as jest.Mock<MigrationReporter>

/**
 * Runs the default set of transforms
 *
 * @param {string} code
 * @param {State} [state]
 * @return {*}
 */
const transform = async (code: string, state?: State) => {
  state = state ?? stateBuilder()

  const transforms = defaultTransformerChain
  return transformRunner(code, state, transforms)
}

/**
 * Runs a single watermark transform
 *
 * @param {string} code
 * @param {State} [state]
 * @return {*}
 */
const watermarkTransform = async (code: string, state?: State) => {
  const filePath = './fake/test.js'
  const isTestFile = filePath.endsWith('.test.js')
  state =
    state ??
    stateBuilder({
      config: {
        filePath,
        isTestFile,
        watermark: '@typescriptify',
        watermarkMessage: `
THIS FILE IS AUTOMATICALLY GENERATED. Do not edit this file.
If you want to manually write flow types for this file,
remove the @typescriptify annotation and this comment block.
`,
        convertJSXSpreads: false,
      },
    })

  const transforms = [watermarkTransformRunner]
  return transformRunner(code, state, transforms)
}

const transformRunner = async (
  code: string,
  state: State,
  transforms: readonly Transformer[]
) => {
  const reporter = new MigrationReporter()
  const file: t.File = recast.parse(code, {
    parser: recastFlowParser,
  })

  await runTransforms(reporter, state, file, transforms)

  return recast.print(file, recastOptions).code
}

const expectMigrationReporterMethodCalled = (
  methodName: keyof MigrationReporter
) => {
  const didCall = MockedMigrationReporter.mock.instances.some((reporter) => {
    return (
      (reporter[methodName] as jest.Mock<MigrationReporter[typeof methodName]>)
        .mock.calls.length >= 1
    )
  })
  expect(didCall).toBe(true)
}

const expectMigrationReporterMethodNotCalled = (
  methodName: keyof MigrationReporter
) => {
  const didCall = MockedMigrationReporter.mock.instances.some((reporter) => {
    return (
      (reporter[methodName] as jest.Mock<MigrationReporter[typeof methodName]>)
        .mock.calls.length >= 1
    )
  })
  expect(didCall).toBe(false)
}

type DeepPartialOverride<T> = {
  [P in keyof T]?: DeepPartialOverride<T[P]>
}

type StateLessConfigurableTypeProvider = DeepPartialOverride<
  Omit<State, 'configurableTypeProvider'>
> &
  Partial<Pick<State, 'configurableTypeProvider'>>

const stateBuilder = (
  stateOverrides: StateLessConfigurableTypeProvider = {}
): State => {
  const filePath = './fake/test.js'
  const isTestFile = filePath.endsWith('.test.js')
  const typeProvider =
    stateOverrides.configurableTypeProvider ??
    new ConfigurableTypeProvider({
      useStrictAnyFunctionType: false,
      useStrictAnyObjectType: false,
    })

  return {
    hasJsx: false,
    usedUtils: false,
    ...stateOverrides,
    config: {
      filePath,
      isTestFile,
      watermark: '',
      watermarkMessage: '',
      convertJSXSpreads: false,
      dropImportExtensions: false,
      keepPrivateTypes: false,
      forceTSX: false,
      disableFlow: false,
      ...stateOverrides.config,
    },
    configurableTypeProvider: typeProvider,
  }
}

type ResultDictionary = { [filename: string]: string }

export function createOutputRecorder(): [
  ResultDictionary,
  (file: SourceFile) => void
] {
  const results: ResultDictionary = {}

  function recordResult(file: SourceFile) {
    results[file.getBaseName()] = file.getFullText()
  }

  return [results, recordResult]
}

const getTestFixState = (argv: FixCommandCliArgs): FixCommandState => {
  const migrationReporter = new MigrationReporter()
  const project = new Project({
    tsConfigFilePath: argv.config,
  })

  return { argv, migrationReporter, project }
}

export {
  transform,
  watermarkTransform,
  expectMigrationReporterMethodCalled,
  expectMigrationReporterMethodNotCalled,
  stateBuilder,
  MockedMigrationReporter,
  getTestFixState,
}
