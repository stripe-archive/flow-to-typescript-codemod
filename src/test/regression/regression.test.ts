import fs from 'fs'
import { anyTypeAnnotation } from '@babel/types'
import * as ts from 'typescript'
import dedent from 'dedent'
import { flowTypeAtPos } from '../../convert/flow/type-at-pos'
import { transform } from '../../convert/utils/testing'

jest.mock('../../convert/flow/type-at-pos')

const mockedFlowTypeAtPos = <jest.MockedFunction<typeof flowTypeAtPos>>(
  flowTypeAtPos
)
mockedFlowTypeAtPos.mockResolvedValue(anyTypeAnnotation())

function generateTypeScriptErrorMessages(received: ts.Diagnostic[]) {
  const errorMessages = received.map((diagnostic) => {
    const { line, character } = ts.getLineAndCharacterOfPosition(
      diagnostic.file!,
      diagnostic.start!
    )
    const errorMessage = ts.flattenDiagnosticMessageText(
      diagnostic.messageText,
      '\n'
    )
    return `${diagnostic.file!.fileName} (${line + 1},${
      character + 1
    }): ${errorMessage}`
  })
  return dedent`Received TypeScript errors:
   ${dedent(
     [...new Set(errorMessages)].reduce(
       (returnMessage, error, n) => `${returnMessage + (n + 1)}. ${error}\n\n`,
       ''
     )
   )}`
}

// Custom error message for comparing diagnostics from TS
expect.extend({
  toHaveNoTypeScriptErrors(received: ts.Diagnostic[]) {
    if (received.length === 0) {
      return {
        message: () => 'Did not receive any TypeScript errors!',
        pass: true,
      }
    } else {
      return {
        message: () => generateTypeScriptErrorMessages(received),
        pass: false,
      }
    }
  },
})

describe('Regression tests', () => {
  test('flow_typescript_differences', async () => {
    const transformedData = await getData(
      `${__dirname}/../test-files/flow_typescript_differences.js`
    )

    expect(transformedData).toMatchSnapshot()

    const transformationResult = compileTypeScriptCode(
      'flow_typescript_differences',
      replaceImports(transformedData),
      ['es2015', 'dom']
    )

    expect(transformationResult.diagnostics).toHaveNoTypeScriptErrors()
    expect(transformationResult.success).toBeTruthy()
  })
})

async function getData(filename: string) {
  const data = fs.readFileSync(filename, {
    encoding: 'utf8',
    flag: 'r',
  })

  const transformedData = await transform(data.toString())

  // Remove flow annotation for cleanliness
  return transformedData.replace(/\/\/ @flow.*\n+/, '')
}

// The in memory typescript compiler can't seem to find imported modules.
// This hack to replace imports with absolute paths seems to work.
// Don't apply this before the snapshot, or your snapshot will have absoulte paths
function replaceImports(data: string) {
  return data
    .replace(
      /flow-to-typescript-codemod/,
      require.resolve('../../../flow.d').replace(/.d.ts/, '')
    )
    .replace(
      /'react'/gi,
      `'${require.resolve(
        '../../../node_modules/@types/react/index.d'
      )}'`.replace(/.d.ts/, '')
    )
}

function compileTypeScriptCode(
  sourceFileName: string,
  code: string,
  libs: string[]
): {
  success: boolean
  diagnostics: ts.Diagnostic[]
} {
  const sourceFileNameWithExtension = `${sourceFileName}.ts`
  const options: ts.CompilerOptions = {
    ...ts.getDefaultCompilerOptions(),
    ...{
      // We set this so we can indicate an error based on whether it emitted or not
      noEmitOnError: true,
      // This makes inputting test samples much easier
      noUnusedLocals: false,
      esModuleInterop: true,
      module: ts.ModuleKind.CommonJS,
    },
  }

  // Real Host is based on the file system
  const fileSystemHost = ts.createCompilerHost(options, true)

  // Create a dummy TS source file based on the input code
  const dummySourceFile = ts.createSourceFile(
    sourceFileNameWithExtension,
    code,
    ts.ScriptTarget.Latest
  )

  // Proxy the functions we care about to allow reading files from memory
  const proxiedMemoryHost: ts.CompilerHost = {
    ...fileSystemHost,
    ...{
      fileExists: (filePath) =>
        filePath === sourceFileNameWithExtension ||
        fileSystemHost.fileExists(filePath),
      getSourceFile: (
        fileName,
        languageVersion,
        onError,
        shouldCreateNewSourceFile
      ) =>
        fileName === sourceFileNameWithExtension
          ? dummySourceFile
          : fileSystemHost.getSourceFile(
              fileName,
              languageVersion,
              onError,
              shouldCreateNewSourceFile
            ),
      readFile: (filePath) =>
        filePath === sourceFileNameWithExtension
          ? code
          : fileSystemHost.readFile(filePath),
      writeFile: () => {
        // Do nothing
      },
    },
  }

  // Load TypeScript Libs
  const rootNames = libs.map((lib) =>
    require.resolve(`typescript/lib/lib.${lib}.d.ts`)
  )
  const program = ts.createProgram(
    rootNames.concat([sourceFileNameWithExtension]),
    options,
    proxiedMemoryHost
  )
  const emitResult = program.emit()
  const diagnostics = ts.getPreEmitDiagnostics(program)
  return {
    success: !emitResult.emitSkipped,
    diagnostics: emitResult.diagnostics.concat(diagnostics),
  }
}
