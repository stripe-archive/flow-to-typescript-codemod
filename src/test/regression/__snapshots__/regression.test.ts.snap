// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Regression tests flow_typescript_differences 1`] = `
"import {Flow} from 'flow-to-typescript-codemod';
// Examples taken from https://github.com/niieani/typescript-vs-flowtype

// bounded polymorphism
function fooGood<T extends {
  x: number;
}>(obj: T): T {
  console.log(Math.abs(obj.x));
  return obj;
}

// Maybe and nullable
let a: string | null | undefined;

// Optional params
function fnOptional(x?: number) {}

// Casting
(1 + 1 as number);

// Exact/partial types
type ExactUser = {
  name: string;
  age: number;
};
type User = {
  name: string;
  age: number;
};
type OptionalUser = Partial<User>;

// Mixed
function stringifyNum(num: number) {
  // Do stuff
}

function stringify(value: unknown) {
  if (typeof value === 'string') {
    return '' + value; // Works!
  }
  if (typeof value === 'number') {
    return stringifyNum(value); // Works!
  }
  return '';
}

// Class types
class Test {}
type TestType = typeof Test;

const instance = new Test();
type TestTypeFromInstance = Flow.Class<typeof instance>;

// Keys
var props = {
  foo: 1,
  bar: 'two',
  baz: 'three',
};

function fnUnion(x?: number) {}

// Type inference
function addOne(a: any) {
  return a + 1;
}
const addOneArrow = (a: any) => a + 1;

const onePlus = [1, 2, 3].map(function addOne(a) {
  return a + 1;
});

const onePlusArrow = [1, 2, 3].map((a) => a + 1);

// Lookup types
type A = {
  thing: string;
};

// when the property is a string constant use $PropertyType (i.e. you know it when typing)
type lookedUpThing = A['thing'];

// when you want the property to be dynamic use $ElementType (since Flow 0.49)
// Fix me
// function getProperty<T : Object, Key : string>(obj: T, key: Key): $ElementType<T, Key> {
//     return obj[key];
// }+

// Type narrowing
// Fix me
// function isNil(value: mixed): boolean %checks {
//   return value == null;
// }

// const thing = null;

// if (!isNil(thing)) {
//   const another = thing.something;
// }

// Call
type Fn1 = <T>(arg1: T) => T;
type E = ReturnType<Fn1>;

declare var e: E; // E is number
(42 as E); // OK

// Mapped types
type InputType = {
  hello: string;
};

type MappedType = Flow.ObjMap<InputType, () => number>;

type FormFieldDef<T> = {
  label: string;
  description?: string;
  default: T | (() => T);
};

type ExtractShape = <T>(arg1: T) => FormFieldDef<T>;

type FormFields<T> = Flow.ObjMap<T, ExtractShape>;

// // Overloading
// Fix me
// declare function add(x: string, y: string): string;
// declare function add(x: number, y: number): number;

declare class Adder {
  add(x: string, y: string): string;
  add(x: number, y: number): number;
}

// Readonly
type B = {
  readonly b: string;
};

let b: B = {b: 'something'};

// Difference
type C = Flow.Diff<{
  a: string;
  b: number;
}, {
  a: string;
}>;

// Rest
type Props = {
  name: string;
  age: number;
};

const propsRest: Props = {name: 'Jon', age: 42};
const {age, ...otherProps} = propsRest;
(otherProps as Partial<Flow.Diff<Props, {
  age: number;
}>>);

// Same
type F = {
  (): string;
};
const f: F = () => 'hello';
const hello: string = f();
"
`;
